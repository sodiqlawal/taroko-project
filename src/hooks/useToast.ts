import { TOAST_OPTION } from "@/constant";
import { ToastContext } from "@/context/ToastContext";
import { useContext } from "react";

export const useToast = () => {
  const { setToast } = useContext(ToastContext);

  const toast = ({
    type,
    message,
  }: {
    type: TOAST_OPTION;
    message: string;
  }) => {
    setToast({ isOpen: true, type, message });

    setTimeout(() => {
      setToast({ isOpen: false, type, message: "" });
    }, 5000);
  };

  return {
    toast,
  };
};
