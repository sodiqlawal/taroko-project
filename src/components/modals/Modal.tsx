import type { FC, PropsWithChildren } from "react";
import { ModalController } from "@/hooks/useModal";
import { GrClose } from "react-icons/gr";
import classNames from "classnames";
import styles from "./modal.module.scss";

const Modal: FC<
    PropsWithChildren<{
        controller: ModalController;
        title?: string;
        className?: string;
    }>
> = ({ controller, children, title, className }) => {
    if (!controller.isOpen) return null;
    return (
        <div className={styles.modal} data-testid="modal">
            <div
                className={classNames(
                    styles.paper,
                    className
                )}
            >
                <div
                    className={styles.close}
                    onClick={controller.close}
                >
                    <GrClose className="text-gray-600" />
                </div>

                <h2>{title}</h2>
                
                <div className={styles.children}>{children}</div>
            </div>
        </div>
    );
};

export default Modal;
