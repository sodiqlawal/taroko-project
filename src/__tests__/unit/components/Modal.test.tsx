import { render, screen } from "@testing-library/react";
import Modal from "@/components/modals/Modal";
import { ModalController } from "@/hooks/useModal";

interface DefaultPropsType {
    controller: ModalController;
}

const defaultProps: DefaultPropsType = {
    controller: {
        open: jest.fn,
        isOpen: false,
        close: jest.fn,
        toggle: jest.fn,
        modalData: null,
    },
};

const setupModal = (props = defaultProps) =>
    render(
        <Modal {...props} />
    );

describe("The <Modal /> component", () => {
    it("should match snapshot", () => {
        const { asFragment } = setupModal();

        expect(asFragment()).toMatchSnapshot()
    });

    it("should not render modal if isOpen is false", async () => {
        setupModal();

        expect(screen.queryByText("modal")).not.toBeInTheDocument();
    });

    it("should render modal if isOpen is true", async () => {
        defaultProps.controller.isOpen = true;

        setupModal();

        expect(screen.getByTestId("modal")).toBeInTheDocument();
    });
});
