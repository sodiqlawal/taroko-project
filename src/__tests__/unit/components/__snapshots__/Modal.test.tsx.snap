// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Modal /> component should match snapshot 1`] = `<DocumentFragment />`;
